x/delegation/keeper/delegation.go

```diff
@@ -136,3 +148,107 @@ func (k *Keeper) UndelegateFrom(ctx sdk.Context, params *delegationtype.Delegati
 	// call the hooks registered by the other modules
 	return k.Hooks().AfterUndelegationStarted(ctx, params.OperatorAddress, delegationtype.GetUndelegationRecordKey(r.BlockNumber, r.LzTxNonce, r.TxHash, r.OperatorAddr))
 }
+
+func (k *Keeper) MarkSelfDelegatedOperator(
+	ctx sdk.Context,
+	clientChainID uint64,
+	operatorAddress sdk.AccAddress,
+	stakerAddress common.Address,
+) error {
+	if !k.operatorKeeper.IsOperator(ctx, operatorAddress) {
+		return delegationtype.ErrOperatorNotExist
+	}
+
+	stakerID, _ := assetstype.GetStakeIDAndAssetID(clientChainID, stakerAddress[:], nil)
+	oldOperator, err := k.GetSelfDelegatedOperator(ctx, stakerID)
+	if err != nil {
+		return err
+	}
+
+	if oldOperator != "" && oldOperator == operatorAddress.String() {
+		return delegationtype.ErrOperatorAlreadyMarked
+	}
+
+	var oldOperatorAccAddr sdk.AccAddress
+	if oldOperator != "" {
+		oldOperatorAccAddr, err = sdk.AccAddressFromBech32(oldOperator)
+		if err != nil {
+			return errorsmod.Wrapf(delegationtype.OperatorAddrIsNotAccAddr, "oldOperator:%s", oldOperator)
+		}
+	}
+
+	opFunc := func(keys *delegationtype.SingleDelegationInfoReq, amounts *delegationtype.DelegationAmounts) error {
+		// decrease the share of old operator
+		if oldOperator != "" && keys.OperatorAddr == oldOperator {
+			err = k.assetsKeeper.UpdateOperatorAssetState(ctx, oldOperatorAccAddr, keys.AssetID, assetstype.DeltaOperatorSingleAsset{
+				OperatorShare: amounts.UndelegatableShare.Neg(),
+			})
+		}
+		// increase the share of new marked operator
+		if keys.OperatorAddr == operatorAddress.String() {
+			err = k.assetsKeeper.UpdateOperatorAssetState(ctx, operatorAddress, keys.AssetID, assetstype.DeltaOperatorSingleAsset{
+				OperatorShare: amounts.UndelegatableShare,
+			})
+		}
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	err = k.IterateDelegationsForStaker(ctx, stakerID, opFunc)
+	if err != nil {
+		return err
+	}
+
```

Comment:
We should verify via signing the ownership of the `stakerID` by the `operatorAddress`

Reply:
The assets are controlled by the staker, so the marking should check the signature of the staker. This function will use the input staker address from the Exocore gateway directly. The signature verification will be performed when processing the EVM transaction. The gateway contract should use the recovered caller address as the input staker address when calling the precompiled contract. This documentation provides more information.
https://www.notion.so/exocore/Marking-self-delegation-and-the-staking-compatibility-for-the-other-non-EVM-client-chains-48cfa40ec5eb41b0a2a484c1bda0f65b?pvs=4